src/NHibernate/Cache/CacheBatcher.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using NHibernate.Cache.Access;
+using NHibernate.Engine;
+using NHibernate.Persister.Collection;
+using NHibernate.Persister.Entity;
+
+namespace NHibernate.Cache
+{
+	/// <summary>
+	/// A batcher for batching operations of <see cref="ICacheConcurrencyStrategy"/>, where the batch size is retrived
+	/// from an <see cref="IEntityPersister"/> or <see cref="ICollectionPersister"/>.
+	/// When a different persister or a different operation is added to the batch, the current batch will be executed.
+	/// </summary>
+	internal partial class CacheBatcher
```

Comment:
I am realizing a bit late this class is maybe mostly not necessary.

- It contains logic as if it could handle different kind of batches, but only put batches currently exist. This logic is likely a remnant from conception where a get batch was planned (or even done then removed).
- Why should put batches be segregated by persisters and restricted by `batch-size`?

`batch-size` is meant for get in my point of view. Limiting the batch size reduces the likeliness of loading unneeded data, since the batched gets are done on the assumption that when accessing an entity or collection, the user will later use the other pending entities or collections. This is of course not guaranteed.

But for put, instead, NHibernate will not put data that do not need to be put. So for me, limiting by batch-size here does not really make sense.

And I guess the segregation by persisters was done for handling batch-size, so if limiting by batch-size should be removed, it should be removed too.

Reply:
> It contains logic as if it could handle different kind of batches, but only put batches currently exist.

This class was mentioned to batch also other operations of `ICacheConcurrencyStrategy` like `Lock` in order to support batching the `EntityUpdateAction` and `EntityDeleteAction` in the future ([related comment](https://github.com/nhibernate/nhibernate-core/pull/1633#discussion_r200837292)). I've implemented the lock batcher which would be used prior updating or deleting entities, but was not included as the scope of this PR was reduced.

> Why should put batches be segregated by persisters and restricted by batch-size

I took a similar approach as `IBatcher` has, where the "batch_size" is used also for inserts, but I agree that we could ignore the batch size for the put operation and rather having it configured by the cache provider if needed. I didn't yet perfom a test that would test if splitting in smaller batches would increase or reduce the overall performance, I will test it and let you know. 
About segregation by persister, it was done to make the batcher simpler so that we can have only one batch active at the given moment, similar to `AbstractBatcher` with the `CurrentCommand` property.

Reply:
Here are the results with different batch-sizes for `PutMany` operation:

Binary serializer:
```
PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 20
Times per iteration 75ms,66ms,74ms,69ms,71ms
Average 71ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 100
Times per iteration 42ms,39ms,44ms,42ms,46ms
Average 42,6ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 500
Times per iteration 40ms,40ms,40ms,38ms,40ms
Average 39,6ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 1000
Times per iteration 39ms,41ms,37ms,38ms,38ms
Average 38,6ms
```

Json serializer:
```
PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 20
Times per iteration 55ms,54ms,58ms,59ms,58ms
Average 56,8ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 100
Times per iteration 33ms,29ms,34ms,32ms,31ms
Average 31,8ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 500
Times per iteration 29ms,26ms,27ms,29ms,26ms
Average 27,4ms

PutMany operation for 1000 keys with region strategy FastRegionStrategy:
Total iterations: 5
Batch size: 1000
Times per iteration 27ms,25ms,24ms,28ms,27ms
Average 26,2ms
```

The tests were run the same way as explained [here](https://github.com/nhibernate/NHibernate-Caches/pull/45#issuecomment-401609796).
Based on the result, we should remove the batch-size for the `Put` operation as you said.

Reply:
Thanks for having benchmarked this!

Reply:
@fredericDelaporte I tried this with below configuration.

I have been download NHibernate master repository, then build.

EntApiDefMap.cs Fluent mapping
``Cache.ReadWrite();
``
For 233 records NHibernate Profiler shows 234 Redis Cache Access
1 for query cache the others are entity cache retrieve.
Why is not returned batched, I do not understand?
``
public EntApiDef Get(string apiCode)
        {
            return ISession.Query<EntApiDef>()
                           .Where(x => x.Code == apiCode)
                           .WithOptions(options =>
                           {
                               options.SetReadOnly(true);
                               options.SetCacheable(true);
                           }).FirstOrDefault();
        }
``

Reply:
This stuff is not released, and no second level cache supporting it either.

Reply:
@gokhanabatay NHibernate Profiler will currently not tell you the actual cache calls when batching is enabled. Instead, use [Redis Monitor](https://redis.io/commands/monitor) to verify how many calls actually happened.

Reply:
@maca88 
I test it with profiler too and debug the code; bathSize equals  1 so it does not fetch many just fetchs single record.
What am i doing wrong.
@fredericDelaporte  I know as you mentioned its not released, I just download Nhibernate master code and debug it.

``
var batchSize = persister.GetBatchSize();
			if (batchSize > 1 && persister.Cache.PreferMultipleGet())
			{
				// The first item in the array is the item that we want to load
				var entityBatch =
					source.PersistenceContext.BatchFetchQueue.GetEntityBatch(persister, @event.EntityId, batchSize, false);
				// Ignore null values as the retrieved batch may contains them when there are not enough
				// uninitialized entities in the queue
				var keys = new List<CacheKey>(batchSize);
``

Reply:
You have to set either a `default_batch_fetch_size` in [configuration](http://nhibernate.info/doc/nhibernate-reference/session-configuration.html#configuration-optional), or set the `batch-size` in the [mappings](http://nhibernate.info/doc/nhibernate-reference/performance.html#performance-fetching-batch).

Reply:
@fredericDelaporte  I get your point now, NHibernate.Cache providers must implement CacheBase abstract class in order to support get many and put many. 
At this point it seems to be, i will use corememorycache as second level cache provider.
for data consistency i will use NH Post Events in order to sync other instances of my application.
So i will ignore sync time as a result it could have delays.